src/queries.ts

```diff
@@ -47,43 +42,93 @@ async function getStakedPSTHolders(): Promise<StakedPSTHolders> {
-		const query = createQuery(minBlock, maxBlock);
+		const query = createQuery(prevBlock + 1, Math.min(maxBlock, trustedHeight));
 		const response = await sendQuery(query);
-		allEdges.push(...response.edges);
-		hasNextPage = response.pageInfo.hasNextPage;
+		if (response.edges.length) {
+			console.log(`Transactions count: ${response.edges.length}`);
+			const mostRecientTransaction = response.edges[response.edges.length - 1];
+			const height = mostRecientTransaction.node.block.height;
+			writeFileSync(gqlResultName(prevBlock + 1, height), JSON.stringify(response.edges));
+			prevBlock = height;
+			allEdges.push(...response.edges);
+			hasNextPage = response.pageInfo.hasNextPage;
+			console.log(`Query has next page: ${hasNextPage}`);
+		} else {
+			console.log(`Ignoring empty GQL response`);
+			hasNextPage = false;
+		}
 	}
 
 	return allEdges;
 }
 
+/**
+ * Runs the given GQL query
+ * @param query the query object
+ * @throws if the GW returns a syntax error
+ * @returns {GQLTransactionsResultInterface} the returned transactions
+ */
 async function sendQuery(query: Query): Promise<GQLTransactionsResultInterface> {
 	// TODO: implement retry here
-	const response = await fetch(GQL_URL, {
-		method: 'POST',
-		headers: {
-			'Accept-Encoding': 'gzip, deflate, br',
-			'Content-Type': 'application/json',
-			Accept: 'application/json',
-			Connection: 'keep-alive',
-			DNT: '1',
-			Origin: GQL_URL
-		},
-		body: JSON.stringify(query)
-	});
-	const JSONBody = await response.json();
-	const errors: { message: string; extensions: { code: string } } = !JSONBody.data && JSONBody.errors;
-	if (errors) {
-		console.log(`Error in query: \n${JSON.stringify(query, null, 4)}`);
-		throw new Error(errors.message);
+	let pendingRetries = MAX_RETRIES;
+	let responseOk: boolean | undefined;
+
+	while (!responseOk && pendingRetries >= 0) {
+		const response = await Axios.request({
+			method: 'POST',
+			url: GQL_URL,
+			headers: {
+				'Accept-Encoding': 'gzip, deflate, br',
+				'Content-Type': 'application/json',
+				Accept: 'application/json',
+				Connection: 'keep-alive',
+				DNT: '1',
+				Origin: GQL_URL
+			},
+			data: JSON.stringify(query)
+		});
+		responseOk = response.status >= 200 && response.status < 300;
+
+		try {
+			const JSONBody = response.data;
+			const errors: { message: string; extensions: { code: string } } = !JSONBody.data && JSONBody.errors;
+			if (errors) {
+				console.log(`Error in query: \n${JSON.stringify(query, null, 4)}`);
+				throw new Error(JSON.stringify(errors));
+			}
+			// console.log(`Query result: ${JSON.stringify(JSONBody, null, '\t')}`);
+			return JSONBody.data.transactions as GQLTransactionsResultInterface;
+		} catch (e) {
+			pendingRetries--;
+			console.log(`Error while querying. Retrying (${pendingRetries}). ${JSON.stringify(e)}`);
+			continue;
+		}
 	}
-	return JSONBody.data.transactions as GQLTransactionsResultInterface;
+	throw new Error(`Retries on the query failed!`);
```

Comment:
Why are we tracking `responseOk` here? By following the code paths, I can't see where it would come into play. If you have a response within that range that throws an error in the try/catch, you would skip all the retries?

I would probably expect it to retry on that type of error regardless of whether the last response code was in an okay range

Generally you'll want a little delay on a network retry mechanism as well. I feel like I just finished this on `GatewayAPI`. Here is the delay function there, which is basically from `axios-retry`:

```ts
	private async exponentialBackOffAfterFailedRequest(retryNumber: number): Promise<void> {
		const delay = Math.pow(2, retryNumber) * this.initialErrorDelayMS;
		await new Promise((res) => setTimeout(res, delay));
	}
```

Reply:
Keep in mind you're tracking `pendingRetries` here instead of `retryNumber`. So you'll need to adjust it a little

Reply:
08f038a958b9db79cb0afba4fe2a4f2aa70c36a0

Reply:
Retry instead of throwing: a8e6e7967397ed989b5cb0e2df24df0a437ba4d4